sw/host/opentitanlib/src/signer/manifest_def.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::signer::manifest::*;
+use crate::util::bigint::fixed_size_bigint;
+use crate::util::num_de::HexEncoded;
+use crate::util::parse_int::ParseInt;
+
+use anyhow::{anyhow, Result};
+use serde::Deserialize;
+use std::convert::{TryFrom, TryInto};
+
+fixed_size_bigint!(ManifestRsa, 3072);
+
+#[derive(Debug, Deserialize)]
+struct ManifestBigInt(Option<HexEncoded<ManifestRsa>>);
+
+#[derive(Debug, Deserialize)]
+struct ManifestSmallInt<T: ParseInt>(Option<HexEncoded<T>>);
+
+/// A macro for wrapping manifest struct definitions that parse from HJSON.
+///
+/// The #[repr(C)] version of `Manifest` can only be built when the fields in `ManifestDef` are
+/// present. This macro sets up the field by field conversion and provides the field names for
+/// purposes of error reporting.
+macro_rules! manifest_def {
+    (struct $name:ident {
+        $($field_name:ident: $field_type:ty,)*
+    }, $out_type:ident) => {
+        #[derive(Deserialize, Debug)]
+        struct $name {
+            $($field_name: $field_type,)*
+        }
+
+        impl ManifestPacked<$out_type> for $name {
+            fn unpack(self, _name: &'static str) -> Result<$out_type> {
+                Ok($out_type {
+                    // Call `unpack()` on each field with the field's name included for use in
+                    // error messages.
+                    $($field_name: self.$field_name.unpack(stringify!($field_name))?.try_into()?,)*
+                })
+            }
+        }
+    }
+}
+
+trait ManifestPacked<T> {
+    /// The default error for missing fields.
+    fn unpack_err(&self, name: &'static str) -> Result<T> {
+        Err(anyhow!("Manifest is missing field {}", name))
+    }
+
+    /// Unpack optional fields in the manifest, and error if the field isn't defined.
+    fn unpack(self, name: &'static str) -> Result<T>;
+}
+
+impl ManifestPacked<ManifestRsa> for ManifestBigInt {
+    fn unpack(self, name: &'static str) -> Result<ManifestRsa> {
+        match self.0 {
+            Some(v) => Ok(v.0),
+            None => self.unpack_err(name),
+        }
+    }
+}
+
+impl<T: ParseInt> ManifestPacked<T> for ManifestSmallInt<T> {
+    fn unpack(self, name: &'static str) -> Result<T> {
+        match self.0 {
+            Some(v) => Ok(v.0),
+            None => self.unpack_err(name),
+        }
+    }
+}
+
+impl<T: ParseInt + Default + Copy> ManifestPacked<[T; 8]> for [ManifestSmallInt<T>; 8] {
+    fn unpack(self, name: &'static str) -> Result<[T; 8]> {
+        let mut res: [T; 8] = [Default::default(); 8];
+        let mut i = 0;
+        for e in self {
+            res[i] = e.unpack(name)?;
+            i += 1;
+        }
```

Comment:
@mcy I was wondering if you had a more idiomatic suggestion for this (ideally one that doesn't require `Copy` or `Default` bounds on `T`)

Reply:
Best I can think of: https://godbolt.org/z/xYMooxfox